### ISSUE TITLE
nomad client doesn't reconnecting to nomad servers

### ISSUE THREAD

**Issue by pznamensky** [hashicorp/nomad#3223]
### Nomad version
Nomad v0.7.0-dev (75221f98082deb89ad4bb5ed13190f7f7fe631f2)

### Operating system and Environment details
Centos 7.4
Linux 4.9

### Issue
Nomad client doesn't exit and doesn't keep trying to connect to nomad servers if there was an error in name resolution at start up.
That's why nomad can't get began works himself and supervisor doesn't restart nomad because it still running.

### Reproduction steps

1. stop nomad client
2. ensure that nomad servers are defined using hostnames
3. disable or break DNS resolver on the nomad client machine
4. start nomad client
5. get DNS resolver back

Expected result:
nomad client exit with status code != 0
or
nomad client keep trying to connect to nomad servers

### Nomad Client logs
```
    nomad: Loaded configuration from /etc/nomad.d/nomad-client.hcl
    nomad: ==> Starting Nomad agent...
    nomad[1854]: client: using state directory /var/lib/nomad/client
    nomad[1854]: client: using alloc directory /var/lib/nomad/alloc
    nomad[1854]: client: built-in fingerprints: [arch cgroup consul cpu host memory network nomad signal storage vault env_aws env_gce]
    nomad[1854]: fingerprint.cgroups: cgroups are available
    nomad[1854]: fingerprint.cpu: frequency: 3499 MHz
    nomad[1854]: fingerprint.cpu: core count: 1
    nomad[1854]: client: fingerprinting consul every 15s
    nomad[1854]: client: fingerprinting cgroup every 15s
    nomad[1854]: fingerprint.network: link speed for eth0 set to 1000
    nomad[1854]: fingerprint.network: Detected interface eth0 with IP: <ip-addr>
    nomad[1854]: client: fingerprinting vault every 15s
    nomad[1854]: fingerprint.env_aws: Error querying AWS Metadata URL, skipping
    nomad[1854]: fingerprint.env_gce: Could not read value for attribute "machine-type"
    nomad[1854]: fingerprint.env_gce: Error querying GCE Metadata URL, skipping
    nomad[1854]: client: applied fingerprints [arch cgroup cpu host memory network nomad signal storage]
    nomad[1854]: driver.docker: using client connection initialized from environment
    nomad[1854]: client: fingerprinting rkt every 15s
    nomad[1854]: driver.exec: exec driver is enabled
    nomad[1854]: client: fingerprinting exec every 15s
    nomad[1854]: client: fingerprinting docker every 15s
    nomad[1854]: client: available drivers [mock_driver docker exec java]
    nomad[1854]: client: ignoring server <nomad-server1>:4647 due to resolution error: lookup <nomad-server1> on 137.0.0.1:53: read udp 192.168.122.234:29380->137.0.0.1:53: i/o timeout
    nomad[1854]: client: ignoring server <nomad-server2>:4647 due to resolution error: lookup <nomad-server2> on 137.0.0.1:53: read udp 192.168.122.234:14956->137.0.0.1:53: i/o timeout
    nomad[1854]: client: ignoring server <nomad-server3>:4647 due to resolution error: lookup <nomad-server3> on 137.0.0.1:53: read udp 192.168.122.234:19397->137.0.0.1:53: i/o timeout
    nomad[1854]: client: ignoring server <nomad-server4>:4647 due to resolution error: lookup <nomad-server4> on 137.0.0.1:53: read udp 192.168.122.234:35692->137.0.0.1:53: i/o timeout
    nomad[1854]: client: None of the configured servers are valid: 4 error(s) occurred:#012#012* lookup <nomad-server1> on 137.0.0.1:53: read udp 192.168.122.234:29380->137.0.0.1:53: i/o timeout#012* lookup <nomad-server2> on 137.0.0.1:53: read udp 192.168.122.234:14956->137.0.0.1:53: i/o timeout#012* lookup <nomad-server3> on 137.0.0.1:53: read udp 192.168.122.234:19397->137.0.0.1:53: i/o timeout#012* lookup <nomad-server4> on 137.0.0.1:53: read udp 192.168.122.234:35692->137.0.0.1:53: i/o timeout
    nomad[1854]: client: registration waiting on servers
    nomad[1854]: client: Node ID "cb2de56d-e30a-6f3d-05a2-b8591a7b985b"
    nomad[1854]: agent: not registering Nomad HTTPS Health Check because it requires Consul>=0.7.2
    nomad: ==> Nomad agent configuration:
    nomad: Client: true
    nomad: Log Level: DEBUG
    nomad: Region: global (DC: staging)
    nomad: Server: false
    nomad: Version: 0.7.0-dev
    nomad: ==> Nomad agent started! Log data will stream in below:
    nomad: 2017/09/15 18:02:07.364925 [INFO] client: using state directory /var/lib/nomad/client
    nomad: 2017/09/15 18:02:07.365232 [INFO] client: using alloc directory /var/lib/nomad/alloc
    nomad: 2017/09/15 18:02:07.366259 [DEBUG] client: built-in fingerprints: [arch cgroup consul cpu host memory network nomad signal storage vault env_aws env_gce]
    nomad: 2017/09/15 18:02:07.366634 [INFO] fingerprint.cgroups: cgroups are available
    nomad: 2017/09/15 18:02:07.367129 [DEBUG] fingerprint.cpu: frequency: 3499 MHz
    nomad: 2017/09/15 18:02:07.367353 [DEBUG] fingerprint.cpu: core count: 1
    nomad: 2017/09/15 18:02:07.368474 [DEBUG] client: fingerprinting consul every 15s
    nomad: 2017/09/15 18:02:07.368903 [DEBUG] client: fingerprinting cgroup every 15s
    nomad: 2017/09/15 18:02:07.369736 [DEBUG] fingerprint.network: link speed for eth0 set to 1000
    nomad: 2017/09/15 18:02:07.369999 [DEBUG] fingerprint.network: Detected interface eth0 with IP: <ip-addr>
    nomad: 2017/09/15 18:02:07.371741 [DEBUG] client: fingerprinting vault every 15s
    nomad: 2017/09/15 18:02:09.371687 [DEBUG] fingerprint.env_aws: Error querying AWS Metadata URL, skipping
    nomad: 2017/09/15 18:02:10.424908 [DEBUG] fingerprint.env_gce: Could not read value for attribute "machine-type"
    nomad: 2017/09/15 18:02:10.424952 [DEBUG] fingerprint.env_gce: Error querying GCE Metadata URL, skipping
    nomad: 2017/09/15 18:02:10.424966 [DEBUG] client: applied fingerprints [arch cgroup cpu host memory network nomad signal storage]
    nomad: 2017/09/15 18:02:10.425034 [DEBUG] driver.docker: using client connection initialized from environment
    nomad: 2017/09/15 18:02:10.425228 [DEBUG] client: fingerprinting rkt every 15s
    nomad: 2017/09/15 18:02:10.428139 [DEBUG] driver.exec: exec driver is enabled
    nomad: 2017/09/15 18:02:10.428506 [DEBUG] client: fingerprinting exec every 15s
    nomad[1854]: consul.sync: failed to update services in Consul: error querying Consul services: Get http://127.0.0.1:8500/v1/agent/services: dial tcp 127.0.0.1:8500: getsockopt: connection refused
    nomad: 2017/09/15 18:02:10.428615 [DEBUG] client: fingerprinting docker every 15s
    nomad: 2017/09/15 18:02:10.476958 [DEBUG] client: available drivers [mock_driver docker exec java]
    nomad: 2017/09/15 18:02:30.478712 [DEBUG] client: ignoring server <nomad-server1>:4647 due to resolution error: lookup <nomad-server1> on 137.0.0.1:53: read udp 192.168.122.234:29380->137.0.0.1:53: i/o timeout
    nomad: 2017/09/15 18:02:50.479725 [DEBUG] client: ignoring server <nomad-server2>:4647 due to resolution error: lookup <nomad-server2> on 137.0.0.1:53: read udp 192.168.122.234:14956->137.0.0.1:53: i/o timeout
    nomad: 2017/09/15 18:03:10.480826 [DEBUG] client: ignoring server <nomad-server3>:4647 due to resolution error: lookup <nomad-server3> on 137.0.0.1:53: read udp 192.168.122.234:19397->137.0.0.1:53: i/o timeout
    nomad: 2017/09/15 18:03:30.481956 [DEBUG] client: ignoring server <nomad-server4>:4647 due to resolution error: lookup <nomad-server4> on 137.0.0.1:53: read udp 192.168.122.234:35692->137.0.0.1:53: i/o timeout
    nomad: 2017/09/15 18:03:30.482032 [WARN] client: None of the configured servers are valid: 4 error(s) occurred:
    nomad: * lookup <nomad-server1> on 137.0.0.1:53: read udp 192.168.122.234:29380->137.0.0.1:53: i/o timeout
    nomad: * lookup <nomad-server2> on 137.0.0.1:53: read udp 192.168.122.234:14956->137.0.0.1:53: i/o timeout
    nomad: * lookup <nomad-server3> on 137.0.0.1:53: read udp 192.168.122.234:19397->137.0.0.1:53: i/o timeout
    nomad: * lookup <nomad-server4> on 137.0.0.1:53: read udp 192.168.122.234:35692->137.0.0.1:53: i/o timeout
    nomad: 2017/09/15 18:03:30.482500 [DEBUG] client: registration waiting on servers
    nomad: 2017/09/15 18:03:30.482731 [INFO] client: Node ID "cb2de56d-e30a-6f3d-05a2-b8591a7b985b"
    nomad: 2017/09/15 18:03:30.482764 [WARN] agent: not registering Nomad HTTPS Health Check because it requires Consul>=0.7.2
    nomad: 2017/09/15 18:03:30.492451 [WARN] consul.sync: failed to update services in Consul: error querying Consul services: Get http://127.0.0.1:8500/v1/agent/services: dial tcp 127.0.0.1:8500: getsockopt: connection refused
    nomad[1854]: client: registration waiting on servers
    nomad: 2017/09/15 18:03:58.832538 [DEBUG] client: registration waiting on servers
    nomad: ==> Failed to check for updates: Get https://checkpoint-api.hashicorp.com/v1/check/nomad?arch=amd64&os=linux&signature=1083fb2c-f810-c6f2-4a3f-6b9153ddb557&version=0.7.0-dev: dial tcp: lookup checkpoint-api.hashicorp.com on 137.0.0.1:53: read udp 192.168.122.234:42390->137.0.0.1:53: i/o timeout
    nomad[1854]: client: registration waiting on servers
    nomad: 2017/09/15 18:04:15.285068 [DEBUG] client: registration waiting on servers
    nomad[1854]: client: registration waiting on servers
    nomad: 2017/09/15 18:04:31.092001 [DEBUG] client: registration waiting on servers
    nomad[1854]: client: registration waiting on servers
    nomad: 2017/09/15 18:04:51.170021 [DEBUG] client: registration waiting on servers
    nomad[1854]: client: registration waiting on servers
    nomad: 2017/09/15 18:05:11.302148 [DEBUG] client: registration waiting on servers
    nomad[1854]: client: registration waiting on servers
    nomad: 2017/09/15 18:05:29.176668 [DEBUG] client: registration waiting on servers
    nomad[1854]: client: registration waiting on servers
    nomad: 2017/09/15 18:05:57.774415 [DEBUG] client: registration waiting on servers
    nomad[1854]: client: registration waiting on servers
    nomad: 2017/09/15 18:06:19.427955 [DEBUG] client: registration waiting on servers
    nomad[1854]: client: registration waiting on servers
    nomad: 2017/09/15 18:06:47.040227 [DEBUG] client: registration waiting on servers
    nomad[1854]: client: registration waiting on servers
    nomad: 2017/09/15 18:07:15.632285 [DEBUG] client: registration waiting on servers
    nomad[1854]: client: registration waiting on servers
    nomad: 2017/09/15 18:07:40.496729 [DEBUG] client: registration waiting on servers
    nomad[1854]: client: registration waiting on servers
    nomad: 2017/09/15 18:07:59.211026 [DEBUG] client: registration waiting on servers
    nomad[1854]: client: registration waiting on servers
    nomad: 2017/09/15 18:08:14.614118 [DEBUG] client: registration waiting on servers
    nomad[1854]: client: registration waiting on servers
```

**Comment by epipho** [comment]
I am seeing what I believe is a variant of this bug.

- Nomad servers are specified using a hostname (A record w/ multiple IPs).
- Take one of the servers offline
- On startup, nomad client resolves ones, if it picks the down server IP it will never re-resolve and thus never connect.

**Comment by schmichael** [comment]
Thanks for the detailed report @epipho and @pznamensky!

**Comment by DrRibosome** [comment]
Also experiencing this

It seems like the issue is that a dns name is only resolved *once* when the nomad client starts. If the nomad server host changes for whatever reason, the nomad client doesnt resolve an ip from the dns entry and continuously tries to connect to the old (now incorrect) ip.

**Comment by ifokeev** [comment]
Same problem here too. I have docker containers with nomad and when masters restart before the clients then clients fail to connect again

**Comment by tgross** [comment]
@DrRibosome and @ifokeev this is a bit of an older issue, so it would help our tracking of the problem if you mentioned which version you were using so we can be sure it's still a problem on more recent versions of Nomad.

> It seems like the issue is that a dns name is only resolved once when the nomad client starts. If the nomad server host changes for whatever reason, the nomad client doesnt resolve an ip from the dns entry and continuously tries to connect to the old (now incorrect) ip.

I took a quick look through this area of the code and that's part of the problem. The Nomad client gets its initial server from the [`server_join`](https://www.nomadproject.io/docs/configuration/server_join#server_join-parameters) stanza that might include the IP, DNS, or a `go-discovery` configuration. But after that it gets updated server information from the servers whenever it sends a [`Node.UpdateStatus`](https://github.com/hashicorp/nomad/blob/v0.12.3/client/client.go#L1791) and that list is always a list of IP addresses.

From the logs in the original issue, it's not clear to me that the IP addresses that the servers are handing back to the clients are routable from the client. If the servers are giving the client private IP addresses, then the client will be able to perform its initial connection (because of the explicit `server_join`) but then never be able to reconnect on failure.

Typically the way you'd solve this sort of problem is by setting the [`advertise.rpc`](https://www.nomadproject.io/docs/configuration#rpc-1) configuration on the servers.

**Comment by ifokeev** [comment]
@tgross I see in the logs "rpc timeout". So is it possible to fix this error? RPC is setting up on the client start. We provide dns name for RPC, but nomad uses internal IP by default

**Comment by tgross** [comment]
Going to close this. If you can reproduce this issue with a current version of Nomad, please open a new issue with the appropriate diagnostic information.

**Comment by ifokeev** [comment]
@tgross could you, please, link the PR? If it's not fixed then it's still present. I couldn't find it in release notes

**Comment by tgross** [comment]
There wasn't an associated PR. We asked some specific questions about the environment and never got a reply. If you are encountering this issue in your current environment, please open a new GitHub issue with the appropriate information.
